### ISSUE TITLE
Harmony error message on 0.16.1393 in Steam Linux

### ISSUE THREAD

**Issue by DanielCoffey** [FluffierThanThou/StackMerger#4]
Rimworld Version : 0.16.1393 rev 539
Platform : Steam
OS : Linux Mint 18.1 Cinnamon

Mod List : Core, Stack Merger (13 Feb 2017 A16)

When Stack Merger is downloaded from the Steam Workshop and is the only mod (right underneath Core), the following error message is seen when RimWorld is opened...

"Could not execute post-long-event action. Exception: System.TypeInitializationException: An exception was thrown by the type initializer for StackMerger.HarmonyPatches ---> System.EntryPointNotFoundException: mono_domain_get"

I have attached the full log file below (I had to append .txt to the file to get GitHub to accept it)...

[rimworld_log.txt](https://github.com/FluffierThanThou/StackMerger/files/804383/rimworld_log.txt)

**Comment by FluffierThanThou** [comment]
Hey Daniel, thanks for taking the time to create an issue :+1: 

Unfortunately, this is a (known) issue with the Harmony library, so there's not much I can do about it right now. I've let the author of Harmony know what's up, he may want to get in touch with you in order to test possible patches. 

I'll try and keep you informed on the status.

**Comment by FluffierThanThou** [comment]
I've just updated Harmony to 1.0.8, which will hopefully solve the issue. Please let me know!

**Comment by DanielCoffey** [comment]
Sadly no change in the error log. Harmony is still throwing that same error. I stopped and started Steam and it went and got the new version of Stack Merger. Should I un-sub and re-sub from the mod instead?

[rimworld_log.txt](https://github.com/FluffierThanThou/StackMerger/files/845725/rimworld_log.txt)

**Comment by FluffierThanThou** [comment]
Ok I've had contact with the Harmony creator, and it turns out he was just waiting for confirmation that it works on linux before releasing. 

You have randomly been selected to be (part of) that confirmation!  Congratulations!

Anyway, please try the version here;
https://github.com/FluffierThanThou/StackMerger/releases/tag/v0.16.0.7

**Comment by DanielCoffey** [comment]
I have just tried downloading that linked version and it won't unpack on Linux. Archive Manager thinks the file is empty even though it is 228Kb. I'll have a look at it in Windows and see if it is confused too.

**Comment by FluffierThanThou** [comment]
ok, that's odd. It works fine on my machine (Win). You could just download
the current repo, it should work just fine as well (it's on the same
version as that pre-release).

**Comment by DanielCoffey** [comment]
It unpacked fine in Windows so I just copied it over to Linux unpacked. I dropped the Stack Merger folder into the Mods folder alongside Core, saw the new mod in the Rimworld Mods list in-game and ticked it. On restart I noted that the original error is gone but I get a new one now...

Could not execute post-long-event action. Exception: System.TypeInitializationException: An exception was thrown by the type initializer for StackMerger.HarmonyPatches ---> System.DllNotFoundException: libmono.so
  at (wrapper managed-to-native) Harmony.ILCopying.MonoInternals:mono_domain_get__L64 ()

[rimworld_log.txt](https://github.com/FluffierThanThou/StackMerger/files/846068/rimworld_log.txt)

**Comment by FluffierThanThou** [comment]
ok, quick update, we (well mostly https://github.com/pardeike) are working on it. Turns out it's really difficult to tell mono where to import dlls from at runtime, and RimWorld puts the required dll at a non-standard location. 

A short-term solution should be to copy/symlink the libmono.so file that should be in the rimworld/data/mono dir to /usr/lib/libmono.so. If you're comfortable with and able to do that, it should work, but we haven't actually tested this, so let us know your results.

**Comment by DanielCoffey** [comment]
Ooh, looks like it loads happily with a symlink as a work-around.

I had to look up how to use Terminal to make symlinks since it involves permissions because /usr/lib/ is owned by root. I used...

cd /usr/lib
sudo ln -s ~/.local/share/Steam/steamapps/common/RimWorld/RimWorldLinux_Data/Mono/x86_64/libmono.so

Here is the latest log file. There are still some complaints from other mods about Mono but nothing from Stack Merger.

[rimworld_log.txt](https://github.com/FluffierThanThou/StackMerger/files/846210/rimworld_log.txt)

I have not had time to try the mod itself as it is late here but please pass my thanks to the Harmony team.

**Comment by FluffierThanThou** [comment]
Cool, let us know your further results :)

**Comment by DanielCoffey** [comment]
I am happy to report that with the symlink above, StackMerger and the new Harmony library works perfectly as expected under Linux. I observed an idle colonist take a couple of loose Simple Meals and drop them onto the other part stack nearby, neatly consolidating the two piles.

Obviously without the symlink, RimWorld throws a wobbler about the missing dll and doesn't load StackMerger.

This was under Linux Mint 18.1 running Kernel 4.4.0-67 with the Steam version of RimWorld.

**Comment by FluffierThanThou** [comment]
good to hear! 

We've been working on a more long-term solution, which should work without the symlink. Would you mind removing the symlink and testing this version?

https://github.com/FluffierThanThou/StackMerger/releases/tag/v0.16.0.8

**Comment by pardeike** [comment]
Yes, I am very much interested if the new version Fluffy is using in the above link works. You may even remove the symlink because I made Harmony work completely without any dll requirements. I want to release as soon as possible and this is the last puzzle piece missing. Cheers!

**Comment by DanielCoffey** [comment]
Gah - timing! Sorry guys but my Linux PC died this afternoon with a cpu_led and error 00 which means either the PSU is dead, the CPU is toast or possibly the motherboard. I am leaning towards the CPU as it was rather overclocked. Dead stable for three years and now just dead.

Tomorrow I have to do a drain and full tear down to check the obvious issues then it will be shopping for a new mobo/cpu/ram plus new rigid pipe bending for water cooling.

If it resurrects then I will of course test the new build otherwise will keep you informed of when I am likely to be back. Till then I am using an iPad and old Mac Mini.

**Comment by pardeike** [comment]
wow. good luck man!

**Comment by FluffierThanThou** [comment]
ouch, good luck with it!

**Comment by DanielCoffey** [comment]
The timing really sucks. I know jumping from a 4770K to a 7700K is lovely but given it needs a new motherboard and ram too it gets pricey. I will keep the old GPU and ssd drives for now but after a tear down and check tomorrow I should be able to get the new kit by Monday. Allowing a few days for pipe bending and leak tests I should be back in business by the end of next week.

Ah well, an excuse for some new toys anyway.

**Comment by FluffierThanThou** [comment]
given that harmony 1.0.9 is now officially out, I've updated. When you've got your system resurrected, let me know if it works :)

**Comment by DanielCoffey** [comment]
Of course. All my new stuff arrives on Saturday but being a custom water system, it will take me a while to get it set up and leak tested. It is an odd combination of old tech (780ti and sata3 ssd) but Kaby Lake 7700k and ddr4 ram. I will have to wait a few months for an opportunity to upgrade the rest.

**Comment by DanielCoffey** [comment]
I am happy to report that the new Steam version of Stack merger works perfectly with the new Harmony library. I deleted the symlink, rebooted, loaded Steam and saw the new version of Stack merger come down. I removed the manual version of the mod and ticked the Steam one, reloaded RimWorld and all was well.

The PC rebuild went fine too - parts arrived at 14:00 yesterday and by 22:00 I had fitted everything, made some new copper tubing runs and passed the leak test. All I had to do this morning was stop the ROG board doing its LED effects, unleash the memory to its full speed and find out why Windows had taken over the boot loader duties instead of Linux. All sorted now, thankfully.
